File: Problem17/main.java
public class main {
    public static void main(String[] args) {
        assert letters(1) == "one".length();
        assert letters(9) == "nine".length();
        assert letters(11) == "eleven".length();
        assert letters(19) == "nineteen".length();
        assert letters(20) == "twenty".length();
        assert letters(21) == "twentyone".length();
        assert letters(29) == "twentynine".length();
        assert letters(101) == "onehundredandone".length();
        assert letters(144) == "onehundredandfortyfour".length();
        assert letters(24) == "twentyfour".length();
        assert letters(990) == "ninehundredandninety".length();
        assert letters(7) == "seven".length();
        assert letters(19) == "nineteen".length();
        assert letters(85) == "eightyfive".length();
        assert letters(200) == "twohundred".length();
        assert letters(1000) == "onethousand".length();
        int sum = 0;
        for (int i = 1; i <= 1000; i++) {
            sum += letters(i);
        }
        System.out.println(sum);
    }

    private static int letters(int i) {
        String ones[] = { "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };
        String tens[] = { "ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen" };
        String decs[] = { "", "", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety" };
        
        if (i == 0) {
            return 0;
        } else if (i < 10) {
            return ones[i - 1].length();
        } else if (i < 20) {
            return tens[i - 10].length();
        } else if (i < 100) {
            return decs[i/10].length() + letters(i - i/10*10);
        } else if (i < 1000) {
            int count = letters(i/100) + "hundred".length();
            if (i%100 == 0) {
                return count;
            }
            return count + "and".length() + letters(i - i/100*100);
        } else {
            return "onethousand".length();
        }
    }
}

File: Problem17/main.hs
main :: IO ()
main = print $ length $ concat $ concatMap words $ map verbalize $ [1 .. 1000]

verbalize :: Int -> [Char]
verbalize n
    | n == 1000 = "one thousand"
    | n >= 100 && hm == 0 = verbalizeDigit hd ++ " hundred"
    | n >= 100 = verbalizeDigit hd ++ " hundred and " ++ verbalize hm
    | n >= 20 && tm == 0 = verbalizeTensDigit td
    | n >= 20 = verbalizeTensDigit td ++ " " ++ verbalize tm
    | n >= 10 = verbalizeTeen tm
    | otherwise = verbalizeDigit n
    where
        (hd, hm) = n `divMod` 100
        (td, tm) = n `divMod` 10
        verbalizeDigit = (!!) (words "zero one two three four five six seven eight nine")
        verbalizeTensDigit = (!!) (words ". . twenty thirty forty fifty sixty seventy eighty ninety")
        verbalizeTeen = (!!) (words "ten eleven twelve thirteen fourteen fifteen sixteen seventeen eighteen nineteen")
